File: ios/Classes/AdvancedBusinessCardReaderPlugin.swift
import Flutter
import UIKit
import MLKitVision
import MLKitTextRecognition
import MLKitTextRecognitionChinese
import MLKitTextRecognitionDevanagari
import MLKitTextRecognitionJapanese
import MLKitTextRecognitionKorean
import MLKitEntityExtraction

public class AdvancedBusinessCardReaderPlugin: NSObject, FlutterPlugin {
  private var recognizerCache: [String: TextRecognizer] = [:]
  private var extractorCache: [String: EntityExtractor] = [:]

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(
      name: "advanced_business_card_reader",
      binaryMessenger: registrar.messenger()
    )
    let instance = AdvancedBusinessCardReaderPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "ocrFromFile":
      handleOcrFromFile(call, result: result)
    case "extractEntities":
      handleExtractEntities(call, result: result)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  public func detachFromEngine(for registrar: FlutterPluginRegistrar) {
    recognizerCache.removeAll()
    extractorCache.removeAll()
  }

  private func handleOcrFromFile(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard
      let args = call.arguments as? [String: Any],
      let path = args["path"] as? String,
      !path.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    else {
      result(
        FlutterError(
          code: "INVALID_ARGUMENT",
          message: "Missing required argument: path",
          details: nil
        )
      )
      return
    }

    guard let image = UIImage(contentsOfFile: path) else {
      result(
        FlutterError(
          code: "OCR_ERROR",
          message: "Failed to load image from file path.",
          details: nil
        )
      )
      return
    }

    let script = normalizeScript((args["script"] as? String) ?? "latin")
    let recognizer = getOrCreateRecognizer(script: script)
    let visionImage = VisionImage(image: image)
    visionImage.orientation = image.imageOrientation

    recognizer.process(visionImage) { recognizedText, error in
      if let error = error {
        result(
          FlutterError(
            code: "OCR_ERROR",
            message: error.localizedDescription,
            details: nil
          )
        )
        return
      }

      guard let recognizedText = recognizedText else {
        result(
          FlutterError(
            code: "OCR_ERROR",
            message: "OCR returned no result.",
            details: nil
          )
        )
        return
      }

      let blocks: [[String: Any]] = recognizedText.blocks.map { block in
        let frame = block.frame
        return [
          "text": block.text,
          "boundingBox": [
            "left": Int(frame.minX.rounded()),
            "top": Int(frame.minY.rounded()),
            "right": Int(frame.maxX.rounded()),
            "bottom": Int(frame.maxY.rounded())
          ]
        ]
      }

      result([
        "text": recognizedText.text.trimmingCharacters(in: .whitespacesAndNewlines),
        "blocks": blocks
      ])
    }
  }

  private func handleExtractEntities(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard
      let args = call.arguments as? [String: Any],
      let rawText = args["text"] as? String
    else {
      result([
        "languageUsed": "en",
        "entities": []
      ])
      return
    }

    let text = rawText.trimmingCharacters(in: .whitespacesAndNewlines)
    if text.isEmpty {
      result([
        "languageUsed": "en",
        "entities": []
      ])
      return
    }

    let requestedLanguage = (args["language"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines)
    let modelIdentifier = resolveModelIdentifier(language: requestedLanguage)
    let languageUsed = modelIdentifier.toLanguageTag()
    let extractor = getOrCreateExtractor(modelIdentifier: modelIdentifier)

    extractor.downloadModelIfNeeded { [weak self] downloadError in
      if downloadError != nil {
        result([
          "languageUsed": languageUsed,
          "entities": []
        ])
        return
      }

      extractor.annotateText(text) { annotations, extractionError in
        if extractionError != nil || annotations == nil {
          result([
            "languageUsed": languageUsed,
            "entities": []
          ])
          return
        }

        let entities = self?.toEntityMaps(annotations: annotations ?? [], text: text) ?? []
        result([
          "languageUsed": languageUsed,
          "entities": entities
        ])
      }
    }
  }

  private func getOrCreateRecognizer(script: String) -> TextRecognizer {
    if let cached = recognizerCache[script] {
      return cached
    }

    let recognizer: TextRecognizer
    switch script {
    case "chinese":
      recognizer = TextRecognizer.textRecognizer(options: ChineseTextRecognizerOptions())
    case "devanagari":
      recognizer = TextRecognizer.textRecognizer(options: DevanagariTextRecognizerOptions())
    case "japanese":
      recognizer = TextRecognizer.textRecognizer(options: JapaneseTextRecognizerOptions())
    case "korean":
      recognizer = TextRecognizer.textRecognizer(options: KoreanTextRecognizerOptions())
    default:
      recognizer = TextRecognizer.textRecognizer(options: TextRecognizerOptions())
    }

    recognizerCache[script] = recognizer
    return recognizer
  }

  private func getOrCreateExtractor(modelIdentifier: EntityExtractionModelIdentifier) -> EntityExtractor {
    let key = modelIdentifier.toLanguageTag()
    if let cached = extractorCache[key] {
      return cached
    }

    let options = EntityExtractorOptions(modelIdentifier: modelIdentifier)
    let extractor = EntityExtractor.entityExtractor(options: options)
    extractorCache[key] = extractor
    return extractor
  }

  private func resolveModelIdentifier(language: String?) -> EntityExtractionModelIdentifier {
    guard let language = language, !language.isEmpty else {
      return .english
    }

    let normalized = language.replacingOccurrences(of: "_", with: "-")
    if let identifier = EntityExtractionModelIdentifier.fromLanguageTag(normalized) {
      return identifier
    }

    if let baseIdentifier = EntityExtractionModelIdentifier.fromLanguageTag(
      normalized.components(separatedBy: "-").first ?? normalized
    ) {
      return baseIdentifier
    }

    return .english
  }

  private func toEntityMaps(annotations: [EntityAnnotation], text: String) -> [[String: Any]] {
    var items: [[String: Any]] = []
    let nsText = text as NSString

    for annotation in annotations {
      let start = max(0, annotation.range.location)
      let end = min(nsText.length, annotation.range.location + annotation.range.length)
      let safeRange = NSRange(location: start, length: max(0, end - start))
      let snippet = nsText.substring(with: safeRange)

      for entity in annotation.entities {
        var mapped: [String: Any] = [
          "type": mapEntityType(entity.entityType),
          "text": snippet,
          "start": safeRange.location,
          "end": safeRange.location + safeRange.length
        ]

        let meta = mapEntityMeta(entity)
        if !meta.isEmpty {
          mapped["meta"] = meta
        }
        items.append(mapped)
      }
    }

    return items
  }

  private func mapEntityType(_ type: EntityType) -> String {
    switch type {
    case .phone:
      return "phone"
    case .email:
      return "email"
    case .URL:
      return "url"
    case .address:
      return "address"
    case .dateTime:
      return "datetime"
    case .flightNumber:
      return "flight_number"
    case .IBAN:
      return "iban"
    case .ISBN:
      return "isbn"
    case .money:
      return "money"
    case .paymentCard:
      return "payment_card"
    case .trackingNumber:
      return "tracking_number"
    default:
      return "unknown"
    }
  }

  private func mapEntityMeta(_ entity: Entity) -> [String: Any] {
    var meta: [String: Any] = [:]

    if entity.entityType == .dateTime, let dateTimeEntity = entity.dateTimeEntity {
      meta["timestampMillis"] = Int(dateTimeEntity.dateTime.timeIntervalSince1970 * 1000.0)
      meta["granularity"] = Int(dateTimeEntity.dateTimeGranularity.rawValue)
    }

    if entity.entityType == .money, let moneyEntity = entity.moneyEntity {
      meta["currency"] = moneyEntity.unnormalizedCurrency
      meta["integerPart"] = Int(moneyEntity.integerPart)
      meta["fractionalPart"] = Int(moneyEntity.fractionalPart)
    }

    return meta
  }

  private func normalizeScript(_ script: String) -> String {
    let normalized = script.lowercased()
    let supported = ["latin", "devanagari", "chinese", "japanese", "korean"]
    return supported.contains(normalized) ? normalized : "latin"
  }
}

File: android/src/main/kotlin/com/nh97/advanced_business_card_reader/AdvancedBusinessCardReaderPlugin.kt
package com.nh97.advanced_business_card_reader

import android.content.Context
import android.net.Uri
import com.google.mlkit.nl.entityextraction.Entity
import com.google.mlkit.nl.entityextraction.EntityAnnotation
import com.google.mlkit.nl.entityextraction.EntityExtraction
import com.google.mlkit.nl.entityextraction.EntityExtractionParams
import com.google.mlkit.nl.entityextraction.EntityExtractor
import com.google.mlkit.nl.entityextraction.EntityExtractorOptions
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.text.TextRecognition
import com.google.mlkit.vision.text.TextRecognizer
import com.google.mlkit.vision.text.chinese.ChineseTextRecognizerOptions
import com.google.mlkit.vision.text.devanagari.DevanagariTextRecognizerOptions
import com.google.mlkit.vision.text.japanese.JapaneseTextRecognizerOptions
import com.google.mlkit.vision.text.korean.KoreanTextRecognizerOptions
import com.google.mlkit.vision.text.latin.TextRecognizerOptions
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.io.File
import java.util.Locale

class AdvancedBusinessCardReaderPlugin : FlutterPlugin, MethodChannel.MethodCallHandler {
    private lateinit var channel: MethodChannel
    private lateinit var context: Context

    private val recognizerCache = mutableMapOf<String, TextRecognizer>()
    private val extractorCache = mutableMapOf<String, EntityExtractor>()

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        channel = MethodChannel(binding.binaryMessenger, CHANNEL_NAME)
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            METHOD_OCR_FROM_FILE -> handleOcrFromFile(call, result)
            METHOD_EXTRACT_ENTITIES -> handleExtractEntities(call, result)
            else -> result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        recognizerCache.values.forEach { recognizer ->
            runCatching { recognizer.close() }
        }
        recognizerCache.clear()

        extractorCache.values.forEach { extractor ->
            runCatching { extractor.close() }
        }
        extractorCache.clear()
    }

    private fun handleOcrFromFile(call: MethodCall, result: MethodChannel.Result) {
        val path = call.argument<String>("path")?.trim()
        if (path.isNullOrEmpty()) {
            result.error("INVALID_ARGUMENT", "Missing required argument: path", null)
            return
        }

        val script = normalizeScript(call.argument<String>("script"))
        val image = try {
            InputImage.fromFilePath(context, Uri.fromFile(File(path)))
        } catch (e: Exception) {
            result.error("OCR_ERROR", e.message ?: "Failed to read image file", null)
            return
        }

        val recognizer = try {
            getOrCreateRecognizer(script)
        } catch (e: Exception) {
            result.error("OCR_ERROR", e.message ?: "Failed to initialize recognizer", null)
            return
        }

        recognizer.process(image)
            .addOnSuccessListener { visionText ->
                val blocks = visionText.textBlocks.map { block ->
                    val box = block.boundingBox
                    mapOf(
                        "text" to block.text,
                        "boundingBox" to if (box == null) {
                            null
                        } else {
                            mapOf(
                                "left" to box.left,
                                "top" to box.top,
                                "right" to box.right,
                                "bottom" to box.bottom,
                            )
                        },
                    )
                }

                result.success(
                    mapOf(
                        "text" to visionText.text.trim(),
                        "blocks" to blocks,
                    ),
                )
            }
            .addOnFailureListener { e ->
                result.error("OCR_ERROR", e.message ?: "OCR processing failed", null)
            }
    }

    private fun handleExtractEntities(call: MethodCall, result: MethodChannel.Result) {
        val text = call.argument<String>("text")?.trim()
        if (text.isNullOrEmpty()) {
            result.success(
                mapOf(
                    "languageUsed" to "en",
                    "entities" to emptyList<Map<String, Any>>(),
                ),
            )
            return
        }

        val requestedLanguage = call.argument<String>("language")?.trim()
        val modelIdentifier = resolveModelIdentifier(requestedLanguage)
        val languageUsed = languageTagForModel(modelIdentifier)
        val extractor = getOrCreateExtractor(modelIdentifier)

        extractor.downloadModelIfNeeded()
            .onSuccessTask {
                extractor.annotate(
                    EntityExtractionParams.Builder(text).build(),
                )
            }
            .addOnSuccessListener { annotations ->
                result.success(
                    mapOf(
                        "languageUsed" to languageUsed,
                        "entities" to toEntityMaps(annotations, text),
                    ),
                )
            }
            .addOnFailureListener {
                // Entity extraction is best-effort only.
                result.success(
                    mapOf(
                        "languageUsed" to languageUsed,
                        "entities" to emptyList<Map<String, Any>>(),
                    ),
                )
            }
    }

    private fun getOrCreateRecognizer(script: String): TextRecognizer {
        return recognizerCache.getOrPut(script) {
            when (script) {
                "chinese" -> TextRecognition.getClient(
                    ChineseTextRecognizerOptions.Builder().build(),
                )

                "devanagari" -> TextRecognition.getClient(
                    DevanagariTextRecognizerOptions.Builder().build(),
                )

                "japanese" -> TextRecognition.getClient(
                    JapaneseTextRecognizerOptions.Builder().build(),
                )

                "korean" -> TextRecognition.getClient(
                    KoreanTextRecognizerOptions.Builder().build(),
                )

                else -> TextRecognition.getClient(TextRecognizerOptions.DEFAULT_OPTIONS)
            }
        }
    }

    private fun getOrCreateExtractor(modelIdentifier: String): EntityExtractor {
        return extractorCache.getOrPut(modelIdentifier) {
            EntityExtraction.getClient(
                EntityExtractorOptions.Builder(modelIdentifier).build(),
            )
        }
    }

    private fun resolveModelIdentifier(language: String?): String {
        val normalized = language
            ?.trim()
            ?.lowercase(Locale.ROOT)
            ?.replace('_', '-')
            ?.takeIf { it.isNotEmpty() }

        if (normalized != null) {
            runCatching { EntityExtractorOptions.fromLanguageTag(normalized) }
                .getOrNull()
                ?.let { return it }

            val base = normalized.substringBefore('-')
            runCatching { EntityExtractorOptions.fromLanguageTag(base) }
                .getOrNull()
                ?.let { return it }
        }

        return EntityExtractorOptions.ENGLISH
    }

    private fun languageTagForModel(modelIdentifier: String): String {
        return runCatching { EntityExtractorOptions.toLanguageTag(modelIdentifier) }
            .getOrDefault("en")
    }

    private fun toEntityMaps(annotations: List<EntityAnnotation>, text: String): List<Map<String, Any>> {
        val entities = mutableListOf<Map<String, Any>>()
        for (annotation in annotations) {
            val start = annotation.start.coerceAtLeast(0)
            val end = annotation.end.coerceIn(start, text.length)
            val snippet = safeSubstring(text, start, end)
            for (entity in annotation.entities) {
                val mapped = mutableMapOf<String, Any>(
                    "type" to mapEntityType(entity.type),
                    "text" to snippet,
                    "start" to start,
                    "end" to end,
                )
                val meta = mapEntityMeta(entity)
                if (meta.isNotEmpty()) {
                    mapped["meta"] = meta
                }
                entities.add(mapped)
            }
        }
        return entities
    }

    private fun safeSubstring(text: String, start: Int, end: Int): String {
        if (text.isEmpty()) return ""
        val safeStart = start.coerceIn(0, text.length)
        val safeEnd = end.coerceIn(safeStart, text.length)
        return text.substring(safeStart, safeEnd)
    }

    private fun mapEntityType(type: Int): String {
        return when (type) {
            Entity.TYPE_PHONE -> "phone"
            Entity.TYPE_EMAIL -> "email"
            Entity.TYPE_URL -> "url"
            Entity.TYPE_ADDRESS -> "address"
            Entity.TYPE_DATE_TIME -> "datetime"
            Entity.TYPE_FLIGHT_NUMBER -> "flight_number"
            Entity.TYPE_IBAN -> "iban"
            Entity.TYPE_ISBN -> "isbn"
            Entity.TYPE_MONEY -> "money"
            Entity.TYPE_PAYMENT_CARD -> "payment_card"
            Entity.TYPE_TRACKING_NUMBER -> "tracking_number"
            else -> "unknown"
        }
    }

    private fun mapEntityMeta(entity: Entity): Map<String, Any> {
        val meta = mutableMapOf<String, Any>()

        when (entity.type) {
            Entity.TYPE_DATE_TIME -> {
                runCatching {
                    val dateTimeEntity = entity.asDateTimeEntity()
                    if (dateTimeEntity != null) {
                        meta["timestampMillis"] = dateTimeEntity.timestampMillis
                        meta["granularity"] = dateTimeEntity.dateTimeGranularity
                    }
                }
            }

            Entity.TYPE_MONEY -> {
                runCatching {
                    val money = entity.asMoneyEntity()
                    if (money != null) {
                        meta["currency"] = money.unnormalizedCurrency
                        meta["integerPart"] = money.integerPart
                        meta["fractionalPart"] = money.fractionalPart
                    }
                }
            }
        }

        return meta
    }

    private fun normalizeScript(script: String?): String {
        return script
            ?.trim()
            ?.lowercase(Locale.ROOT)
            ?.takeIf { it in SUPPORTED_SCRIPTS }
            ?: "latin"
    }

    companion object {
        private const val CHANNEL_NAME = "advanced_business_card_reader"
        private const val METHOD_OCR_FROM_FILE = "ocrFromFile"
        private const val METHOD_EXTRACT_ENTITIES = "extractEntities"

        private val SUPPORTED_SCRIPTS = setOf(
            "latin",
            "devanagari",
            "chinese",
            "japanese",
            "korean",
        )
    }
}
